core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

```diff
@@ -398,11 +415,14 @@ public void insertRecord(
       spill();
     }
 
+    // Check if a new pointer array has to be allocated both before and after a new page allocated,
+    // to make sure there is a pointer array even if spilling is triggered.
     growPointerArrayIfNecessary();
     int uaoSize = UnsafeAlignedOffset.getUaoSize();
     // Need 4 or 8 bytes to store the record length.
     final int required = length + uaoSize;
     acquireNewPageIfNecessary(required);
+    growPointerArrayIfNecessary();
```

Comment:
`acquireNewPageIfNecessary` and  `growPointerArrayIfNecessary` go hand in hand, maybe we should combine them in a single method?

What happens if first `acquireNewPageIfNecessary` and then `growPointerArrayIfNecessary` spills? Will there be a page we can write to?

Reply:
> `acquireNewPageIfNecessary` and `growPointerArrayIfNecessary` go hand in hand, maybe we should combine them in a single method?

Added `allocateMemoryForRecordIfNecessary`.

> What happens if first `acquireNewPageIfNecessary` and then `growPointerArrayIfNecessary` spills? Will there be a page we can write to?

That won't be an issue. `spill()` prevents memory from being freed if no records have been inserted yet (which is the case after a spill). I added a comment to clarify this but it is a bit wordy. Let me know if you see a way to simplify the comment.